### PR TITLE
Decimal implementation

### PR DESCRIPTION
**Pull request by kempeng**
second try

### CODE REVIEW

File: exchanges/anx/anx_wrapper.go [hunk 1]
```diff
@@ -80,57 +80,60 @@ func (a *ANX) UpdateTicker(p pair.CurrencyPair, assetType string) (ticker.Price,
...
 			return tickerPrice, err
 		}
 	} else {
-		tickerPrice.Bid = 0
+		tickerPrice.Bid = decimal.Zero
 	}
 
 	if tick.Data.Low.Value != "" {
-		tickerPrice.Low, err = strconv.ParseFloat(tick.Data.Low.Value, 64)
+		//tickerPrice.Low, err = strconv.ParseFloat(tick.Data.Low.Value, 64)
```
Comment (shazbert) [inline exchanges/anx/anx_wrapper.go]: Please remove comment

File: exchanges/orderbook/orderbook.go [hunk 2]
```diff
@@ -46,24 +47,28 @@ type Orderbook struct {
...
-func (o *Base) CalculateTotalBids() (float64, float64) {
-	amountCollated := float64(0)
-	total := float64(0)
+func (o *Base) CalculateTotalBids() (decimal.Decimal, decimal.Decimal) {
+	amountCollated := decimal.Zero
+	total := decimal.Zero
 	for _, x := range o.Bids {
-		amountCollated += x.Amount
-		total += x.Amount * x.Price
+		//amountCollated += x.Amount
```
Comment (shazbert) [inline exchanges/orderbook/orderbook.go]: Please remove comment

File: exchanges/orderbook/orderbook.go [hunk 3]
```diff
@@ -46,24 +47,28 @@ type Orderbook struct {
...
+		amountCollated = amountCollated.Add(x.Amount)
+		//total += x.Amount * x.Price
```
Comment (shazbert) [inline exchanges/orderbook/orderbook.go]: Please remove comment

File: exchanges/orderbook/orderbook.go [hunk 4]
```diff
@@ -46,24 +47,28 @@ type Orderbook struct {
...
-func (o *Base) CalculateTotalAsks() (float64, float64) {
-	amountCollated := float64(0)
-	total := float64(0)
+func (o *Base) CalculateTotalAsks() (decimal.Decimal, decimal.Decimal) {
+	amountCollated := decimal.Zero
+	total := decimal.Zero
 	for _, x := range o.Asks {
-		amountCollated += x.Amount
-		total += x.Amount * x.Price
+		//amountCollated += x.Amount
```
Comment (shazbert) [inline exchanges/orderbook/orderbook.go]: Please remove comment

File: exchanges/orderbook/orderbook.go [hunk 5]
```diff
@@ -46,24 +47,28 @@ type Orderbook struct {
...
+		amountCollated = amountCollated.Add(x.Amount)
+		//total += x.Amount * x.Price
```
Comment (shazbert) [inline exchanges/orderbook/orderbook.go]: Please remove comment

File: exchanges/orderbook/orderbook_test.go [hunk 6]
```diff
@@ -4,21 +4,28 @@ import (
...
 		Pair:         currency,
 		CurrencyPair: currency.Pair().String(),
-		Bids:         []Item{{Price: 100, Amount: 10}},
+		Bids:         []Item{{Price: Hundred, Amount: Ten}},
 		LastUpdated:  time.Now(),
 	}
 
 	a, b := base.CalculateTotalBids()
-	if a != 10 && b != 1000 {
+	//if a != 10 && b != 1000 {
```
Comment (shazbert) [inline exchanges/orderbook/orderbook_test.go]: Please remove comment

File: exchanges/orderbook/orderbook_test.go [hunk 7]
```diff
@@ -29,12 +36,13 @@ func TestCalculateTotaAsks(t *testing.T) {
...
 		Pair:         currency,
 		CurrencyPair: currency.Pair().String(),
-		Asks:         []Item{{Price: 100, Amount: 10}},
+		Asks:         []Item{{Price: Hundred, Amount: Ten}},
 		LastUpdated:  time.Now(),
 	}
 
 	a, b := base.CalculateTotalAsks()
-	if a != 10 && b != 1000 {
+	//if a != 10 && b != 1000 {
```
Comment (shazbert) [inline exchanges/orderbook/orderbook_test.go]: Please remove comment

File: exchanges/anx/anx_wrapper.go (hunk 1 above)
Comment (kempeng) [reply exchanges/anx/anx_wrapper.go]: will do

File: exchanges/orderbook/orderbook.go (hunk 2 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook.go]: will do

File: exchanges/orderbook/orderbook_test.go (hunk 6 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook_test.go]: np!

File: exchanges/orderbook/orderbook.go (hunk 5 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook.go]: np

File: exchanges/orderbook/orderbook.go (hunk 4 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook.go]: np

File: exchanges/orderbook/orderbook.go (hunk 3 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook.go]: np

File: exchanges/orderbook/orderbook_test.go (hunk 7 above)
Comment (kempeng) [reply exchanges/orderbook/orderbook_test.go]: np

**Comment by shazbert** [comment]
@kempeng @thrasher- Kempeng did some bench-marking of the decimal package as discussed earlier, I  think that due the rounding effects of float64's that it should still be implemented still. 
```
on an i7 gen8, with 1.000.000 iterations
Work By Kempeng

018/08/20 22:54:40 decimal.NewFromString: 1.1420403s, 499810660.0217192028184311837
2018/08/20 22:54:41 strconv.ParseFloat: 131.6235ms, 499810660.021743
2018/08/20 22:54:45 Orderbook Decimal: 3.0318566s
2018/08/20 22:54:46 Orderbook Float64: 5.9713ms
```

If anybody has any objections or specialist advice, please let us know. There are still some minor fixes/alterations planned for this branch.

File: common/decimal_helpers.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	return f
+}
+
+// NotEqual returns true when d is not equal to d2
+func NotEqual(d, d2 decimal.Decimal) bool {
+	return !d.Equal(d2)
+}
+
+//NewFromInt returns a decimal with the value of int v
+func NewFromInt(v int) decimal.Decimal {
```
Comment (shazbert) [inline common/decimal_helpers.go]: Change naming convention either decimal.NewFromInt() or common.NewDecimalFromtInt(101) to be a bit more specific across this package.

**Comment by shazbert** [comment]
tACK

**Comment by shazbert** [comment]
@thrasher- Helper functions will be RM'd when upstream updates are merged.

**Comment by kempeng** [comment]
Hi @shazbert , anything I need to do to fix these failed checks?

File: exchanges/okex/okex_types.go [hunk 9]
```diff
@@ -1,19 +1,22 @@
 package okex
 
-import "encoding/json"
+import {
```
[the remark this answers is not in the document]
Comment (kempeng) [reply exchanges/okex/okex_types.go]: thanks, clear. ANything I need to do to correct this?

**Comment by shazbert** [comment]
@kempeng This has been put on the back burner for the moment. Apologies for not commenting sooner. Due to the severe performance hit we might branch this off so people can choose it and have a faster version as base.

**Comment by thrasher-** [comment]
Closed due to this PR being stale, however we still aim to make a branch using the decimal package in the future but there's more critical things to work on currently

**Pull request closed without merging**